### PR TITLE
Feat improve entity deduplication with faiss louvain detection and llm

### PR DESCRIPTION
**Pull request by Naseem77**

### CODE REVIEW

**Comment by Naseem77** [comment]
@copilot full review

**Review by galshubeli — COMMENTED** [review]
## Review

The three-tier architecture (exact match → hard embedding merge → LLM verification) is well-designed. Embedding cache in `ctx.metadata`, Union-Find for transitive closure, and neighbor context in the LLM prompt are all good choices.

However, there are several issues that need addressing before merge. See inline comments.

File: graphrag_sdk/pyproject.toml [hunk 1]
```diff
@@ -28,6 +28,8 @@ dependencies = [
     "python-dotenv>=1.0",
     "tiktoken>=0.5",
     "gliner>=0.2",
+    "faiss-cpu>=1.7",
```
Comment (galshubeli) [inline graphrag_sdk/pyproject.toml:31]: `faiss-cpu` was added but is never used — FAISS was replaced with numpy cosine similarity (`mat_normed @ mat_normed.T`) in a later commit due to OpenMP thread-pool conflicts. This is a ~100MB dependency that adds install friction for zero benefit. Please remove it.

File: graphrag_sdk/pyproject.toml [hunk 2]
```diff
@@ -28,6 +28,8 @@ dependencies = [
...
+    "networkx>=3.0",
```
Comment (galshubeli) [inline graphrag_sdk/pyproject.toml:32]: `networkx` is a heavy core dependency added only for `louvain_communities()` in one strategy. We'd prefer not to add it.

Simpler alternatives that need no external lib:
1. **Skip community detection entirely** — send all ambiguous pairs to LLM directly. `max_llm_pairs` already caps cost, and the whole point of LLM verification is handling the ambiguous zone correctly. Auto-merging intra-community pairs without LLM defeats that purpose.
2. **Connected components** — trivial with the Union-Find already in the code. Same grouping idea, no networkx.
3. **Simple threshold split** — pairs above `(hard + soft) / 2` auto-merge, below goes to LLM.

Option 1 is simplest and safest. Please remove `networkx` and the Louvain logic.

File: graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/llm_verified_resolution.py [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+        embedder: Embedder | None = None,
+        *,
+        hard_threshold: float = 0.95,
+        soft_threshold: float = 0.80,
+        max_llm_pairs: int = 500,
+        max_llm_concurrency: int | None = None,
+        force_summary_threshold: int = 3,
+        max_summary_tokens: int = 500,
+        ann_top_k: int = 50,
+    ) -> None:
```
Comment (galshubeli) [inline graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/llm_verified_resolution.py:117]: `ann_top_k` is accepted and stored but never used anywhere. It was meant for FAISS HNSW `top_k` search, which was replaced with brute-force numpy. Please remove from both `LLMVerifiedResolution` and `SemanticResolution`.

File: graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/llm_verified_resolution.py [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+        if hard_threshold <= soft_threshold:
+            raise ValueError(
+                f"hard_threshold ({hard_threshold}) must be > soft_threshold ({soft_threshold})"
+            )
+        self.llm = llm
+        self.embedder = embedder
+        self.hard_threshold = hard_threshold
+        self.soft_threshold = soft_threshold
+        self.max_llm_pairs = max_llm_pairs
+        self.max_llm_concurrency = max_llm_concurrency
```
Comment (galshubeli) [inline graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/llm_verified_resolution.py:127]: This is a 506-line file with complex logic (embedding, Union-Find, Louvain community detection, LLM batch verification) and zero test coverage. At minimum needs tests for:
1. Hard-threshold merges (no LLM called)
2. Soft-zone pairs trigger LLM, merge only on YES
3. Relationship remapping/dedup after merges
4. Edge cases: single node, no embedder, LLM failure

File: graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/llm_verified_resolution.py [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+                node_to_comm: dict[int, int] = {}
+                for cid, members in enumerate(communities):
+                    for m in members:
+                        node_to_comm[m] = cid
+
+                # Intra-community pairs → hard merge (no LLM needed)
+                for gi, gj, _ in ambiguous_pairs:
+                    # Use -1 and -2 as distinct defaults so unindexed nodes go to LLM (safe)
+                    if node_to_comm.get(gi, -1) == node_to_comm.get(gj, -2):
+                        union(gi, gj)
```
Comment (galshubeli) [inline graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/llm_verified_resolution.py:412]: Auto-merging intra-community pairs without LLM verification is risky. Louvain communities can be large and heterogeneous — entities that are "close" in embedding space are not necessarily the same real-world entity. This could incorrectly merge distinct entities that happen to cluster together.

This is the ambiguous zone by definition — if we were confident about these pairs, they'd be above `hard_threshold`. Sending them all to LLM is safer.

File: graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/llm_verified_resolution.py [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+                    parent[py] = px
+
+            hard_pairs: list[tuple[int, int]] = []
+            ambiguous_pairs: list[tuple[int, int, float]] = []
+
+            # Brute-force cosine similarity via numpy (avoids FAISS OpenMP
+            # thread-pool conflicts with PyTorch on macOS and Windows).
+            sim_matrix = mat_normed @ mat_normed.T  # (n, n), rows are unit-normed
+            for i in range(n_nodes):
+                for j in range(i + 1, n_nodes):
```
Comment (galshubeli) [inline graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/llm_verified_resolution.py:375]: The numpy matrix multiply is fast, but the Python double for-loop iterating over all pairs is O(N²) and will be slow for labels with hundreds of entities. Use vectorized indexing instead:
```python
rows, cols = np.where(np.triu(sim_matrix, k=1) >= self.hard_threshold)
hard_pairs = list(zip(rows.tolist(), cols.tolist()))

rows, cols = np.where(
    (np.triu(sim_matrix, k=1) >= self.soft_threshold)
    & (np.triu(sim_matrix, k=1) < self.hard_threshold)
)
ambiguous_pairs = [(r, c, float(sim_matrix[r, c])) for r, c in zip(rows.tolist(), cols.tolist())]
```
Same issue exists in `semantic_resolution.py`.

File: graphrag_sdk/README.md [hunk 7]
```diff
@@ -166,6 +177,8 @@ Every algorithmic concern is a swappable strategy behind an abstract base class:
 | **Loading** | `LoaderStrategy` | `TextLoader`, `PdfLoader` | Auto-detect by file extension |
 | **Chunking** | `ChunkingStrategy` | `FixedSizeChunking`, `SentenceTokenCapChunking`, `ContextualChunking`, `CallableChunking` | `FixedSizeChunking` (1000 chars, 100 overlap) |
 | **Extraction** | `ExtractionStrategy` | `GraphExtraction` (GLiNER2 + LLM) | GraphExtraction |
+| **Resolution** | `ResolutionStrategy` | `ExactMatchResolution`, `DescriptionMergeResolution`, `SemanticResolution`, `LLMVerifiedResolution` | ExactMatch |
```
Comment (galshubeli) [inline graphrag_sdk/README.md:180]: The strategy table now has duplicate Resolution and Retrieval rows (the old ones at lines 183-184 were not removed). Also the benchmark section below has the same duplication. Please remove the old rows.

File: graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/semantic_resolution.py [hunk 8]
```diff
@@ -229,12 +231,44 @@ async def _fuzzy_merge(
...
+                    miss_vecs = await _asyncio.wait_for(
+                        self.embedder.aembed_documents(miss_names),
+                        timeout=20.0,
+                    )
+                    for node, vec in zip(miss_nodes, miss_vecs):
+                        if vec:
+                            emb_cache[node.id] = vec
+                    logger.info(
+                        "SemanticResolution: embeddings done for '%s' (%d/%d cached)",
+                        label,
```
Comment (galshubeli) [inline graphrag_sdk/src/graphrag_sdk/ingestion/resolution_strategies/semantic_resolution.py:257]: The hardcoded 20s timeout and `logger.info` calls with full node name lists look like debugging leftovers. `logger.info` will be noisy in production. Either remove the verbose logging or downgrade to `logger.debug`. The 20s timeout should be configurable or removed — embedding latency varies widely by provider and batch size.

File: graphrag_sdk/pyproject.toml (hunk 2 above)
Comment (galshubeli) [reply graphrag_sdk/pyproject.toml:32]: try to find way to do louvain without installing networkx
